File: py_anc/algorithms/adfxlms.py
from __future__ import annotations

from typing import Dict

import numpy as np


def adfxlms(params: Dict):
    """Distributed ADFxLMS algorithm."""
    time = params["time"]
    rir_manager = params["rir_manager"]
    network = params["network"]
    l = int(params["L"])
    x = np.asarray(params["reference_signal"], dtype=float)
    d = np.asarray(params["desired_signal"], dtype=float)

    key_ref_mics = list(rir_manager.reference_microphones.keys())
    key_sec_spks = list(rir_manager.secondary_speakers.keys())
    key_err_mics = list(rir_manager.error_microphones.keys())

    ref_idx = {rid: i for i, rid in enumerate(key_ref_mics)}
    sec_idx = {sid: i for i, sid in enumerate(key_sec_spks)}
    err_idx = {eid: i for i, eid in enumerate(key_err_mics)}

    num_ref_mics = len(key_ref_mics)
    num_sec_spks = len(key_sec_spks)
    num_err_mics = len(key_err_mics)

    if x.shape[1] != num_ref_mics:
        raise ValueError(
            f"reference_signal columns ({x.shape[1]}) must match number of reference microphones ({num_ref_mics})."
        )

    n_samples = len(time)

    max_ls_hat = 0
    for sec_id in key_sec_spks:
        ls_hat = len(rir_manager.get_secondary_rir(sec_id, key_err_mics[0]))
        if ls_hat > max_ls_hat:
            max_ls_hat = ls_hat

    x_taps = np.zeros((max(l, max_ls_hat), num_ref_mics), dtype=float)
    e = np.zeros((n_samples, num_err_mics), dtype=float)

    y_taps = []
    for sec_id in key_sec_spks:
        sec_len = len(rir_manager.get_secondary_rir(sec_id, key_err_mics[0]))
        y_taps.append(np.zeros(sec_len, dtype=float))

    for node in network.nodes.values():
        if node.ref_mic_id not in ref_idx:
            raise ValueError(f"Node {node.node_id} has invalid ref_mic_id={node.ref_mic_id}")
        node.init(l)

    print("开始ADFxLMS仿真...")
    for n in range(n_samples):
        x_taps = np.vstack([x[n : n + 1, :], x_taps[:-1, :]])

        for node in network.nodes.values():
            ref_col = ref_idx[node.ref_mic_id]
            self_col = node.neighbor_ids.index(node.node_id)
            y = float(node.phi[:, self_col] @ x_taps[:l, ref_col])
            k = sec_idx[node.sec_spk_id]
            y_taps[k] = np.concatenate(([y], y_taps[k][:-1]))

        for m, err_id in enumerate(key_err_mics):
            yf = 0.0
            for k, sec_id in enumerate(key_sec_spks):
                s = rir_manager.get_secondary_rir(sec_id, err_id)
                ls = len(s)
                yf += float(np.dot(s, y_taps[k][:ls]))
            e[n, m] = d[n, m] + yf

        for node in network.nodes.values():
            ref_col = ref_idx[node.ref_mic_id]
            xf = np.zeros(len(node.neighbor_ids), dtype=float)
            for idx, neighbor_id in enumerate(node.neighbor_ids):
                neighbor = network.nodes[neighbor_id]
                s_hat = rir_manager.get_secondary_rir(neighbor.sec_spk_id, node.err_mic_id)
                ls_hat = len(s_hat)
                xf[idx] = float(s_hat @ x_taps[:ls_hat, ref_col])
            node.xf_taps = np.vstack([xf.reshape(1, -1), node.xf_taps[:-1, :]])

        for node in network.nodes.values():
            node.psi = node.phi - node.step_size * e[n, err_idx[node.err_mic_id]] * node.xf_taps

        for node in network.nodes.values():
            new_phi = np.zeros_like(node.phi)
            for idx, neighbor_id in enumerate(node.neighbor_ids):
                neighbor = network.nodes[neighbor_id]
                com_neighbor_ids = [cid for cid in node.neighbor_ids if cid in neighbor.neighbor_ids]
                if not com_neighbor_ids:
                    continue
                for com_neighbor_id in com_neighbor_ids:
                    com_neighbor = network.nodes[com_neighbor_id]
                    local_idx = com_neighbor.neighbor_ids.index(neighbor_id)
                    new_phi[:, idx] += com_neighbor.psi[:, local_idx] / len(com_neighbor_ids)
            node.phi = new_phi

    filter_coeffs = {}
    for node in network.nodes.values():
        self_col = node.neighbor_ids.index(node.node_id)
        filter_coeffs[node.sec_spk_id] = node.phi[:, self_col].copy()

    return {"err_hist": e, "filter_coeffs": filter_coeffs}

File: +algorithms/ADFxLMS.m
function results = ADFxLMS(params)
% ADFxLMS 分布式主动噪声控制FxLMS算法
%
% 输入:
%   params: 包含所有仿真参数的结构体。
%       - params.time: 离散时间轴
%       - params.rirManager: RIRManager对象，用于获取脉冲响应
%       - params.network: 通讯网络拓扑结构
%       - params.L: 控制滤波器的长度
%       - params.referenceSignal: 参考信号
%       - params.desiredSignal: 期望信号
%
% 输出:
%   results: 包含仿真结果的结构体。
%       - results.err_hist: 麦克风处的误差信号历史
%       - results.filter_coeffs: 各节点最终的控制器滤波器系数
%% 1. 解包参数
time            = params.time;
rirManager      = params.rirManager;
network         = params.network;
L               = params.L;
x               = params.referenceSignal;
d               = params.desiredSignal;

% 从rirManager获取参数
keyRefMics = keys(rirManager.ReferenceMicrophones);
keySecSpks = keys(rirManager.SecondarySpeakers);
keyErrMics = keys(rirManager.ErrorMicrophones);

numRefMics      = numEntries(rirManager.ReferenceMicrophones);
numSecSpks      = numEntries(rirManager.SecondarySpeakers);
numErrMics      = numEntries(rirManager.ErrorMicrophones);

if size(x, 2) ~= numRefMics
    error('referenceSignal columns (%d) must match number of reference microphones (%d).', size(x, 2), numRefMics);
end

nSamples = length(time);

%% 2. 初始化
max_Ls_hat = 0;
for i = keySecSpks'
    Ls_hat = length(rirManager.getSecondaryRIR(i, keyErrMics(1)));
    if Ls_hat > max_Ls_hat
        max_Ls_hat = Ls_hat;
    end
end

x_taps = zeros(max([L, max_Ls_hat]), numRefMics);

e = zeros(nSamples, numErrMics); % 误差信号

y_taps = cell(numSecSpks);   % 控制信号
for k = 1:numSecSpks
    y_taps{k} = zeros(length(rirManager.getSecondaryRIR(keySecSpks(k), keyErrMics(1))), 1);
end

for keyNode = keys(network.Nodes)'
    node = network.Nodes(keyNode);
    node.init(L);
end

%% 3. 主循环
disp('开始ADFxLMS仿真...');
for n = 1:nSamples
    % 3.1. 更新参考信号状态向量 (全局)
    x_taps = [x(n, :); x_taps(1:end-1, :)];

    % 3.2. 生成控制信号 y(n) (分布式)
    for keyNode = keys(network.Nodes)'
        node = network.Nodes(keyNode);
        y = node.Phi(:, node.NeighborIds == node.Id)' * x_taps(1:L, keyRefMics == node.RefMicId);
        y_taps{keySecSpks == node.SecSpkId} = [y; y_taps{keySecSpks == node.SecSpkId}(1:end-1)];
    end

    % 3.3. 计算误差信号 e(n) (全局)
    for m = 1:numErrMics
        yf = 0;
        for k = 1:numSecSpks
            S = rirManager.getSecondaryRIR(keySecSpks(k), keyErrMics(m));
            Ls = length(S);
            yf = yf + S * y_taps{k}(1:Ls);
        end
        e(n, m) = d(n, m) + yf;
    end

    % 3.4. 更新节点信息 (分布式)
    % 3.4.1 滤波参考信号 x_filtered(n)
    for keyNode = keys(network.Nodes)'
        node = network.Nodes(keyNode);
        xf = zeros(1, numel(node.NeighborIds));
        for idx = 1:numel(node.NeighborIds)
            neighbor = network.Nodes(node.NeighborIds(idx));
            S_hat = rirManager.getSecondaryRIR(neighbor.SecSpkId, node.ErrMicId);
            Ls_hat = length(S_hat);
            xf(1, idx) = S_hat * x_taps(1:Ls_hat, keyRefMics == node.RefMicId);
        end
        node.xf_taps = [xf; node.xf_taps(1:end-1, :)];
    end
    % 3.4.2 更新Psi
    for keyNode = keys(network.Nodes)'
        node = network.Nodes(keyNode);
        node.Psi = node.Phi - node.StepSize * e(n, keyErrMics == node.ErrMicId) * node.xf_taps;
    end
    % 3.4.3 更新Phi
    for keyNode = keys(network.Nodes)'
        node = network.Nodes(keyNode);
        node.Phi = zeros(L, numel(node.NeighborIds));
        for idx = 1:numel(node.NeighborIds)
            neighborId = node.NeighborIds(idx);
            neighbor = network.Nodes(neighborId);
            comNeighborIds = intersect(node.NeighborIds, neighbor.NeighborIds);
            for comNeighborId = comNeighborIds
                comNeighbor = network.Nodes(comNeighborId);
                node.Phi(:, idx) = node.Phi(:, idx) + comNeighbor.Psi(:, comNeighbor.NeighborIds == neighborId) / numel(comNeighborIds);
            end
        end
    end
end

%% 4. 打包结果
filter_coeffs = dictionary;
for keyNode = keys(network.Nodes)'
    node = network.Nodes(keyNode);
    w = node.Phi(:, node.NeighborIds == node.Id);
    filter_coeffs(node.SecSpkId) = {w};
end
results.err_hist      = e;
results.filter_coeffs = filter_coeffs;
end